File: sdk/go/registration/pending.go
package registration

import "sort"

// pendingChange 表示同一 Registration 所有连续待处理事件的完整逻辑效果。
// baseRevision 是合并 Update 可安全应用到的本地版本；repair 表示必须回读权威 Hash。
type pendingChange struct {
	event          registrationEvent
	baseRevision   uint64
	latestRevision uint64
	repair         bool
}

// pendingChanges 为每个 UUID 最多保留一个合并项，并同时限制条目数和估算字节数。
type pendingChanges struct {
	entries    map[string]pendingChange
	bytes      int
	maxEntries int
	maxBytes   int
}

// newPendingChanges 创建一个空合并缓冲区；maxEntries 与 maxBytes 均由 Selector 配置提供。
func newPendingChanges(maxEntries int, maxBytes int) pendingChanges {
	return pendingChanges{
		entries:    make(map[string]pendingChange),
		maxEntries: maxEntries,
		maxBytes:   maxBytes,
	}
}

// add 接管 event 的 map 与字节切片，并按 UUID 合并连续变化。
// 订阅读取者交付后不再保留 event；容量拒绝时，接近字节上限的路径保证原缓冲区不被部分修改。
func (pending *pendingChanges) add(event registrationEvent) error {
	current, exists := pending.entries[event.uuid]
	previousBytes := 0
	if exists {
		previousBytes = pendingChangeSize(current)
	}

	var next pendingChange
	var err error
	if exists {
		// 合并结果最多增长一个 incoming 的估算大小。普通路径直接修改已拥有 map；
		// 接近上限时先复制，使最终容量检查失败仍保持原队列事务性。
		incomingBytes := registrationEventSize(event)
		copyCurrent := pending.bytes > pending.maxBytes-incomingBytes
		next, err = mergePendingChange(current, event, copyCurrent)
	} else {
		next = initialPendingChange(event)
	}
	if err != nil {
		return err
	}

	nextBytes := pending.bytes - previousBytes + pendingChangeSize(next)
	if !exists && len(pending.entries) >= pending.maxEntries {
		return protocolError(codeCapacity, "selector_event_entries", 0)
	}
	if nextBytes > pending.maxBytes {
		return protocolError(codeCapacity, "selector_event_bytes", 0)
	}
	pending.entries[event.uuid] = next
	pending.bytes = nextBytes
	return nil
}

// drain 按 UUID 排序返回全部合并项，并原地清空缓冲区供下一轮复用。
// 空缓冲区返回 nil；排序保证修复和测试顺序确定。
func (pending *pendingChanges) drain() []pendingChange {
	if len(pending.entries) == 0 {
		return nil
	}
	uuids := make([]string, 0, len(pending.entries))
	for uuid := range pending.entries {
		uuids = append(uuids, uuid)
	}
	sort.Strings(uuids)
	changes := make([]pendingChange, 0, len(uuids))
	for _, uuid := range uuids {
		changes = append(changes, pending.entries[uuid])
	}
	clear(pending.entries)
	pending.bytes = 0
	return changes
}

// initialPendingChange 从第一条事件建立合并状态，并推导 Update/Renew 的基准版本。
func initialPendingChange(event registrationEvent) pendingChange {
	change := pendingChange{event: event, latestRevision: event.revision}
	if event.kind == "update" {
		change.baseRevision = event.revision - 1
	} else if event.kind == "renew" {
		change.baseRevision = event.revision
	}
	return change
}

// mergePendingChange 把 incoming 合并到 current，并验证生命周期与版本连续性。
// copyCurrent 指示修改前是否深拷贝 current 的字段，以保证容量拒绝时不改变原状态。
func mergePendingChange(current pendingChange, incoming registrationEvent, copyCurrent bool) (pendingChange, error) {
	if incoming.kind == "unregister" {
		return initialPendingChange(incoming), nil
	}
	if current.event.kind == "unregister" {
		return pendingChange{}, protocolError(codeTransition, "@uuid", 0)
	}
	// 一旦进入 repair，只有足够新的完整 Register 能重新建立可直接应用的基线。
	if current.repair {
		if incoming.kind == "register" && incoming.revision >= current.latestRevision {
			return initialPendingChange(incoming), nil
		}
		if incoming.revision > current.latestRevision {
			current.latestRevision = incoming.revision
			current.event.revision = incoming.revision
		}
		if incoming.timestamp > current.event.timestamp {
			current.event.timestamp = incoming.timestamp
		}
		return current, nil
	}

	if incoming.kind == "register" {
		if incoming.revision < current.latestRevision {
			return current, nil
		}
		next := initialPendingChange(incoming)
		if incoming.revision == current.latestRevision {
			if current.event.kind == "register" && !sameRegistrationContent(current.event, incoming) {
				return repairPendingChange(current, incoming), nil
			}
			if current.event.timestamp > next.event.timestamp {
				next.event.timestamp = current.event.timestamp
			}
		}
		return next, nil
	}

	switch incoming.kind {
	case "update":
		return mergePendingUpdate(current, incoming, copyCurrent), nil
	case "renew":
		return mergePendingRenew(current, incoming), nil
	default:
		return pendingChange{}, protocolError(codeInvalid, "&kind", 0)
	}
}

// mergePendingUpdate 合并一条新 Update；版本重复冲突或间隙会退化为定向修复标记。
func mergePendingUpdate(current pendingChange, incoming registrationEvent, copyCurrent bool) pendingChange {
	if incoming.revision <= current.latestRevision {
		if incoming.revision == current.latestRevision && current.event.kind == "update" &&
			(!sameOptionalVersion(current.event, incoming) || !fieldsEqual(current.event.data, incoming.data)) {
			return repairPendingChange(current, incoming)
		}
		return current
	}
	if incoming.revision != current.latestRevision+1 {
		return repairPendingChange(current, incoming)
	}

	// 完整 Register 可在原地吸收连续字段 Patch，前提是 Patch 不引入未知 Data 字段。
	if current.event.kind == "register" {
		next := current
		if copyCurrent {
			next.event = cloneRegistrationEvent(current.event)
		}
		for name, value := range incoming.data {
			if _, exists := next.event.data[name]; !exists {
				return repairPendingChange(current, incoming)
			}
			next.event.data[name] = value
		}
		next.event.revision = incoming.revision
		next.event.timestamp = max(current.event.timestamp, incoming.timestamp)
		if incoming.hasVersion {
			next.event.version = incoming.version
			next.event.hasVersion = true
		}
		next.latestRevision = incoming.revision
		return next
	}

	// Renew 后的首个 Update 替换事件主体但保留旧基准；连续 Update 则按字段覆盖。
	next := current
	if current.event.kind == "renew" {
		next.event = incoming
		next.baseRevision = current.baseRevision
	} else {
		if copyCurrent {
			next.event = cloneRegistrationEvent(current.event)
		}
		if next.event.data == nil {
			next.event.data = make(fields, len(incoming.data))
		}
		for name, value := range incoming.data {
			next.event.data[name] = value
		}
		if incoming.hasVersion {
			next.event.version = incoming.version
			next.event.hasVersion = true
		}
		next.event.revision = incoming.revision
	}
	next.event.timestamp = max(current.event.timestamp, incoming.timestamp)
	next.latestRevision = incoming.revision
	return next
}

// mergePendingRenew 合并同版本续期，只提升时间戳；跨版本续期要求权威修复。
func mergePendingRenew(current pendingChange, incoming registrationEvent) pendingChange {
	if incoming.revision < current.latestRevision {
		return current
	}
	if incoming.revision > current.latestRevision {
		return repairPendingChange(current, incoming)
	}
	if incoming.timestamp > current.event.timestamp {
		current.event.timestamp = incoming.timestamp
	}
	return current
}

// repairPendingChange 把无法安全合并的序列压缩为一个有界定向修复请求。
// 它保留观察到的最大版本和时间戳，帮助后续完整 Register 判断是否足够新。
func repairPendingChange(current pendingChange, incoming registrationEvent) pendingChange {
	latest := max(current.latestRevision, incoming.revision)
	timestamp := max(current.event.timestamp, incoming.timestamp)
	return pendingChange{
		event: registrationEvent{
			kind:      "repair",
			uuid:      incoming.uuid,
			revision:  latest,
			timestamp: timestamp,
		},
		latestRevision: latest,
		repair:         true,
	}
}

// cloneRegistrationEvent 深拷贝 Attr/Data，其余标量按值复制。
func cloneRegistrationEvent(event registrationEvent) registrationEvent {
	event.attr = cloneFields(event.attr)
	event.data = cloneFields(event.data)
	return event
}

// sameRegistrationContent 比较完整 Register 的 TTL、Version、Attr 和 Data，不比较时间戳。
func sameRegistrationContent(left registrationEvent, right registrationEvent) bool {
	return left.ttl == right.ttl && left.version == right.version &&
		fieldsEqual(left.attr, right.attr) && fieldsEqual(left.data, right.data)
}

// sameOptionalVersion 比较 Update 是否都省略 Version，或都携带相同 Version。
func sameOptionalVersion(left registrationEvent, right registrationEvent) bool {
	return left.hasVersion == right.hasVersion && (!left.hasVersion || left.version == right.version)
}

// pendingChangeSize 返回一个合并项的保守内存估算，用于本地缓冲容量控制。
func pendingChangeSize(change pendingChange) int {
	return registrationEventSize(change.event)
}

// registrationEventSize 估算事件持有的字符串、字段 map 和字节切片内存。
// bookkeeping 是每条记录的固定保守开销，不用于线协议长度计算。
func registrationEventSize(event registrationEvent) int {
	const bookkeeping = 128
	size := bookkeeping + len(event.uuid) + len(event.kind)
	for name, value := range event.attr {
		size += 16 + len(name) + len(value)
	}
	for name, value := range event.data {
		size += 16 + len(name) + len(value)
	}
	return size
}

File: sdk/rust/src/registration/pending.rs
use std::collections::BTreeMap;

use super::event::RegistrationEvent;
use crate::error::{Code, Error, Result};

#[derive(Clone, Debug)]
pub(crate) struct PendingChange {
    pub event: RegistrationEvent,
    pub base_revision: u64,
    pub latest_revision: u64,
    pub repair: bool,
}

pub(crate) struct PendingChanges {
    entries: BTreeMap<String, PendingChange>,
    bytes: usize,
    max_entries: usize,
    max_bytes: usize,
}

impl PendingChanges {
    /// 创建同时受 `max_entries` 和 `max_bytes` 约束的空事件合并器。
    pub(crate) fn new(max_entries: usize, max_bytes: usize) -> Self {
        Self {
            entries: BTreeMap::new(),
            bytes: 0,
            max_entries,
            max_bytes,
        }
    }

    /// 接管并按 UUID 合并一条 `event`。
    ///
    /// 每个 UUID 最多保留一个逻辑变化；连续 Update/Renew 原地合并，间隙退化为定向 repair。
    /// 数量和字节拒绝是事务性的：失败保留此前已接纳状态。
    pub(crate) fn add(&mut self, event: RegistrationEvent) -> Result<()> {
        let Some(current) = self.entries.get(&event.uuid) else {
            if self.entries.len() >= self.max_entries {
                return Err(Error::field(Code::Capacity, "selector_event_entries"));
            }
            let next = initial_pending_change(event);
            let next_bytes = self.bytes.saturating_add(pending_change_size(&next));
            if next_bytes > self.max_bytes {
                return Err(Error::field(Code::Capacity, "selector_event_bytes"));
            }
            self.entries.insert(next.event.uuid.clone(), next);
            self.bytes = next_bytes;
            return Ok(());
        };

        let previous_bytes = pending_change_size(current);
        let can_mutate = self.bytes.saturating_add(registration_event_size(&event)) <= self.max_bytes;
        if can_mutate {
            let Some(current) = self.entries.get_mut(&event.uuid) else {
                return Err(Error::field(Code::Corrupt, "selector_pending"));
            };
            merge_pending_change(current, event)?;
            self.bytes = self.bytes.saturating_sub(previous_bytes).saturating_add(pending_change_size(current));
            return Ok(());
        }

        // 临近容量上限时先在克隆上合并，拒绝不会破坏此前已接纳条目。
        let mut next = current.clone();
        merge_pending_change(&mut next, event)?;
        let next_bytes = self.bytes.saturating_sub(previous_bytes).saturating_add(pending_change_size(&next));
        if next_bytes > self.max_bytes {
            return Err(Error::field(Code::Capacity, "selector_event_bytes"));
        }
        let Some(current) = self.entries.get_mut(&next.event.uuid) else {
            return Err(Error::field(Code::Corrupt, "selector_pending"));
        };
        *current = next;
        self.bytes = next_bytes;
        Ok(())
    }

    /// 按 UUID 确定顺序取出全部 PendingChange，并原地重置字节计数与 map。
    pub(crate) fn drain(&mut self) -> Vec<PendingChange> {
        self.bytes = 0;
        std::mem::take(&mut self.entries).into_values().collect()
    }
}

/// 从首条 `event` 推导合并基准、最新 revision 与正常状态。
///
/// Update 基准为 revision-1，Renew 基准为同 revision，Register/Unregister 不依赖本地基准。
fn initial_pending_change(event: RegistrationEvent) -> PendingChange {
    let base_revision = match event.kind.as_str() {
        "update" => event.revision.saturating_sub(1),
        "renew" => event.revision,
        _ => 0,
    };
    PendingChange {
        latest_revision: event.revision,
        event,
        base_revision,
        repair: false,
    }
}

/// 把 `incoming` 的逻辑效果合并到同 UUID 的 `current`。
///
/// 生命周期、同 revision 内容冲突或 revision 间隙会拒绝/转为 repair；Unregister 具有终止优先级。
fn merge_pending_change(current: &mut PendingChange, incoming: RegistrationEvent) -> Result<()> {
    if incoming.kind == "unregister" {
        *current = initial_pending_change(incoming);
        return Ok(());
    }
    if current.event.kind == "unregister" {
        return Err(Error::field(Code::Transition, "@uuid"));
    }
    // repair 状态只跟踪最新 revision/timestamp，直到一条足够新的完整 Register 重建基准。
    if current.repair {
        if incoming.kind == "register" && incoming.revision >= current.latest_revision {
            *current = initial_pending_change(incoming);
            return Ok(());
        }
        if incoming.revision > current.latest_revision {
            current.latest_revision = incoming.revision;
            current.event.revision = incoming.revision;
        }
        current.event.timestamp = current.event.timestamp.max(incoming.timestamp);
        return Ok(());
    }

    if incoming.kind == "register" {
        if incoming.revision < current.latest_revision {
            return Ok(());
        }
        let mut next = initial_pending_change(incoming);
        if next.latest_revision == current.latest_revision {
            if current.event.kind == "register" && !same_registration_content(&current.event, &next.event) {
                *current = repair_pending_change(current, &next.event);
                return Ok(());
            }
            next.event.timestamp = next.event.timestamp.max(current.event.timestamp);
        }
        *current = next;
        return Ok(());
    }

    match incoming.kind.as_str() {
        "update" => {
            merge_pending_update(current, incoming);
            Ok(())
        }
        "renew" => {
            merge_pending_renew(current, &incoming);
            Ok(())
        }
        _ => Err(Error::field(Code::Invalid, "&kind")),
    }
}

/// 合并一条 Update。
///
/// 连续 revision 才可应用；完整 Register 基准原地覆盖已知 Data，Update/Renew 基准压缩为一个增量。
fn merge_pending_update(current: &mut PendingChange, incoming: RegistrationEvent) {
    if incoming.revision <= current.latest_revision {
        if incoming.revision == current.latest_revision
            && current.event.kind == "update"
            && (!same_optional_version(&current.event, &incoming) || current.event.data != incoming.data)
        {
            *current = repair_pending_change(current, &incoming);
        }
        return;
    }
    if incoming.revision != current.latest_revision.saturating_add(1) {
        *current = repair_pending_change(current, &incoming);
        return;
    }

    // 完整 Register 已包含所有固定 Data 字段；出现新字段说明本地结构不可信，必须 repair。
    if current.event.kind == "register" {
        if incoming.data.keys().any(|name| !current.event.data.contains_key(name)) {
            *current = repair_pending_change(current, &incoming);
            return;
        }
        for (name, value) in incoming.data {
            if let Some(field) = current.event.data.get_mut(&name) {
                *field = value;
            }
        }
        current.event.revision = incoming.revision;
        current.event.timestamp = current.event.timestamp.max(incoming.timestamp);
        if incoming.has_version {
            current.event.version = incoming.version;
            current.event.has_version = true;
        }
        current.latest_revision = incoming.revision;
        return;
    }

    let previous_timestamp = current.event.timestamp;
    if current.event.kind == "renew" {
        current.event = incoming;
    } else {
        current.event.data.extend(incoming.data);
        if incoming.has_version {
            current.event.version = incoming.version;
            current.event.has_version = true;
        }
        current.event.revision = incoming.revision;
    }
    current.event.timestamp = previous_timestamp.max(current.event.timestamp);
    current.latest_revision = current.event.revision;
}

/// 合并一条同 revision Renew，只提升 timestamp；跨 revision Renew 转为 repair。
fn merge_pending_renew(current: &mut PendingChange, incoming: &RegistrationEvent) {
    if incoming.revision < current.latest_revision {
        return;
    }
    if incoming.revision > current.latest_revision {
        *current = repair_pending_change(current, incoming);
        return;
    }
    current.event.timestamp = current.event.timestamp.max(incoming.timestamp);
}

/// 把无法安全合并的 `current`/`incoming` 压缩为一个有界定向权威读取标记。
fn repair_pending_change(current: &PendingChange, incoming: &RegistrationEvent) -> PendingChange {
    let latest_revision = current.latest_revision.max(incoming.revision);
    PendingChange {
        event: RegistrationEvent {
            kind: "repair".to_owned(),
            uuid: incoming.uuid.clone(),
            revision: latest_revision,
            timestamp: current.event.timestamp.max(incoming.timestamp),
            ttl: 0,
            version: 0,
            has_version: false,
            attr: Default::default(),
            data: Default::default(),
        },
        base_revision: 0,
        latest_revision,
        repair: true,
    }
}

/// 比较两条完整 Register 的 TTL、Version、Attr 和 Data，不比较时间戳与 revision。
fn same_registration_content(left: &RegistrationEvent, right: &RegistrationEvent) -> bool {
    left.ttl == right.ttl && left.version == right.version && left.attr == right.attr && left.data == right.data
}

/// 比较两条 Update 是否都省略 Version，或都携带相同 Version。
fn same_optional_version(left: &RegistrationEvent, right: &RegistrationEvent) -> bool {
    left.has_version == right.has_version && (!left.has_version || left.version == right.version)
}

/// 返回 `change` 当前事件的保守内存字节估算。
fn pending_change_size(change: &PendingChange) -> usize {
    registration_event_size(&change.event)
}

/// 估算 `event` 的 UUID、kind、字段名/值和 map 条目管理开销。
///
/// 常量 bookkeeping 是保守预算而非 Rust 分配器精确统计，用于稳定地限制合并缓存。
fn registration_event_size(event: &RegistrationEvent) -> usize {
    const BOOKKEEPING: usize = 128;
    BOOKKEEPING
        + event.uuid.len()
        + event.kind.len()
        + event.attr.iter().map(|(name, value)| 16 + name.len() + value.len()).sum::<usize>()
        + event.data.iter().map(|(name, value)| 16 + name.len() + value.len()).sum::<usize>()
}

#[cfg(test)]
#[path = "../../tests/internal/registration/pending.rs"]
mod tests;
